src/pe/resource.rs

```diff
@@ -0,0 +1,1681 @@
+    pub fn file_version(&self) -> Option<String> {
+        self.file_version.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::internal_name`] slice into a [`String`].
+    pub fn internal_name(&self) -> Option<String> {
+        self.internal_name.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::legal_copyright`] slice into a [`String`].
+    pub fn legal_copyright(&self) -> Option<String> {
+        self.legal_copyright.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::legal_trademarks`] slice into a [`String`].
+    pub fn legal_trademarks(&self) -> Option<String> {
+        self.legal_trademarks.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::original_filename`] slice into a [`String`].
+    pub fn original_filename(&self) -> Option<String> {
+        self.original_filename.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::private_build`] slice into a [`String`].
+    pub fn private_build(&self) -> Option<String> {
+        self.private_build.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::product_name`] slice into a [`String`].
+    pub fn product_name(&self) -> Option<String> {
+        self.product_name.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::product_version`] slice into a [`String`].
+    pub fn product_version(&self) -> Option<String> {
+        self.product_version.map(|x| to_utf16_string(x))
+    }
+
+    /// Stringize the [`StringFileInfo::special_build`] slice into a [`String`].
+    pub fn special_build(&self) -> Option<String> {
+        self.special_build.map(|x| to_utf16_string(x))
+    }
+}
+
+/// Represents a version information
+#[derive(Copy, Clone)]
+pub struct VersionInfo<'a> {
+    /// Raw data of entire [`RT_VERSION`] area.
+    pub data: &'a [u8],
+    /// Fixed file information.
+    pub fixed_info: Option<VsFixedFileInfo>,
+    /// Dynamic key-value file information.
+    pub string_info: StringFileInfo<'a>,
+}
+
+impl fmt::Debug for VersionInfo<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("VersionInfo")
+            .field(
+                "data",
+                &format_args!("{:02x?} ({} bytes)", &self.data, self.data.len()),
+            )
+            .field("fixed_info", &self.fixed_info)
+            .field("string_info", &self.string_info)
+            .finish()
+    }
+}
+
+impl<'a> VersionInfo<'a> {
+    pub fn parse(
+        pe: &'a [u8],
+        bytes: &'a [u8],
+        it: ResourceEntryIterator<'a>,
+        sections: &[section_table::SectionTable],
+        file_alignment: u32,
+        opts: &options::ParseOptions,
+    ) -> error::Result<Option<Self>> {
+        let it = it.collect::<Result<Vec<_>, _>>()?;
+        let it = it.iter().find(|e| e.id() == Some(RT_VERSION));
```

Comment:
iiuc, you are finding the element that matches RT_VERSION; you shouldn't have to collect to do this, which is an extra allocation

Reply:
It is still be able to express like following but I don't like that kind of ugly expression:

```rs
    ) -> error::Result<Option<Self>> {
        let mut maybe_entry = None;
        for x in it {
            if x?.id() == Some(RT_VERSION) {
                entry = Some(x?);
                break;
            }
        }

        if let Some(entry) = maybe_entry {
```

This is semantically _not_ equivalent since it hides the error in while converting `Result<T,E>` into `Option<T>` I really don't like:

```rs
        if let Some(entry) = it
            .filter_map(Result::ok)
            .find(|x| x.id() == Some(RT_VERSION))
```

So that I concluded to following implementation:

```rs
impl<'a> ResourceEntryIterator<'a> {
    /// Find the resource entry by its resource ID.
    pub fn find_by_id(&self, id: u16) -> error::Result<Option<ResourceEntry>> {
        self.map(|x| {
            x.and_then(|x| {
                if x.id() == Some(id) {
                    Ok(Some(x))
                } else {
                    Ok(None)
                }
            })
        })
        .find_map(Result::transpose)
        .transpose()
    }
}
```

```rs
    ) -> error::Result<Option<Self>> {
        if let Some(entry) = it.find_by_id(RT_VERSION)? {
```

This also provides shortcut for consumers as `find_by_id` is `pub`.
What do you think about this?